### ISSUE TITLE
multi class oaa progressive loss

### ISSUE THREAD

**Issue by BrianMiner** [JohnLangford/vowpal_wabbit#566]
Before looking at trying to produce a reproducible example (since currently this issue is showing up experimenting on a large kaggle data set) I wanted to check on the logic being employed and make sure there is a good chance the issue I am seeing is really an issue,

The data set is multi-class (9 classes). I compiled the current VW today from github. When I build model with 

vw.exe -d train.vw --oaa 9 -b 24 --loss_function logistic -f  VW.mod

The progressive loss is exceptionally good (0.26). Should this represent the multi-clas log loss?

When predicting a test set using raw_predictions and passing through  a sigmoid I find that the predictions (probabilities) are either all equal or one class will be essentially 1 and the other classes essentially zero. The "hand computed" multi class log loss ends up being quite horrible,

vw.exe -d test.vw -t -i  modVW.mod -r raw_pred.txt 

Is this an issue already known - I did find some discussions of issues with oaa raw_prediction but sounded like it was recently patched.

**Comment by JohnLangford** [comment]
This:

vw.exe -d train.vw --oaa 9 -b 24 --loss_function logistic -f VW.mod

should produce a multiclass loss.

-John

**Comment by martinpopel** [comment]
> Should this represent the multi-clas log loss?

No. VW currently cannot report [multi-class log loss](https://www.kaggle.com/wiki/MultiClassLogLoss).
In `--oaa`, VW reports the 0/1 loss, i.e. 1-accuracy. (Is this what John calls "multiclass loss"?)
I agree it would be very useful if VW could report **both** losses.

**Comment by JohnLangford** [comment]
> Should this represent the multi-clas log loss?

No. VW currently cannot report [multi-class log loss](https://www.kaggle.com/wiki/MultiClassLogLoss).
In `--oaa`, VW reports the 0/1 loss, i.e. 1-accuracy. (Is this what John calls "multiclass loss"?)
I agree it would be very useful if VW could report **both** losses.

Martin

----- Original Message -----

**Comment by BrianMiner** [comment]
Thanks for the info Martin. Yes it would be highly useful. The progressive loss in my case was displayed as 0.26 and then calculated outside of VW, the multi-class log loss was 2.6! I went from figuring out how to accept my kaggle prize to wondering what the heck happened :)

**Comment by martinpopel** [comment]
In binary classification (`vw --loss_function=logistic`), you can choose which loss will be reported: logistic loss (by default) or zero-one (`--binary`).
In multiclass classification, there is no such option, zero-one loss is reported always (although the reduced binary task is optimized using the loss function you have specified with `--loss_function`).
In cost-sensitive multiclass, the reported loss is based on the costs you have specified (so it can be zero-one, but it cannot be multiclass-log-loss).

**Comment by JohnLangford** [comment]
In what scenario do you care about multiclass logistic loss?  I thought
this was always a proxy loss.

-John

**Comment by martinpopel** [comment]
> In what scenario do you care about multiclass logistic loss? 

Let's say the output from VW classifier is reranked by another classifier.
(Example: machine translation, VW is used for the translation model, but we want to rerank it with the language model.)

So I don't care so much if the top-predicted class is the correct one,
but I care if the probabilities assigned to the classes for each example by VW are reasonable.
Multiclass logistic loss somehow tells me this,
although it is not the ideal (intrinsic) evaluation measure for my setting.

For example, I care more about the probabilities of the top-ranked classes
than probabilities of the classes which have no/small chance to be reranked to the 1st position.

I am not aware of any better evaluation measure for my setting
(I've tried precision@5 and some other information extraction measures),
so I use the multiclass logistic loss anyway.

In my experiments, --loss_function=hinge gave me the best zero-one loss,
but --loss_function=logistic gave me much better multiclass-logistic loss
and the final extrinsic evaluation (BLEU score).

Martin

----- Original Message -----

**Comment by MLWave** [comment]
I am experimenting with stacked generalization, feature-weighted linear stacking and distilled/dark knowledge. I currently use logloss for this. Like Martin Popel I care less about correctness of the top prediction ("the object is 93% a car") than I care about the overall correctness of all classes ("the object looks most like a car, a bit like a fire truck and not at all like a carrot"). Multiclass Logistic Loss would be a nice-to-have feature for this.

But we all want to use it for Kaggle competitions :).

**Comment by JohnLangford** [comment]
This:
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/multiclass.cc#L105

is the point in the code where multiclass loss is calculated.   Do one of
you want to tweak it for multiclass log loss?

-John

**Comment by martinpopel** [comment]
Would it be OK to add a column to the VW progress report (so it can report **both** 0/1 and multiclass log loss)?

**Comment by JohnLangford** [comment]
I'd rather not---real estate seems pretty precious.    Why not just
configure?

We could report both in the final printout, I suppose.

-John

**Comment by JohnLangford** [comment]
Ping?

**Comment by martinpopel** [comment]
[sorry for one-month-long ping and a bit too long reply]
Adding  multi-class log loss is not as easy as I thought. The reporting is done in multiclass.cc, which is shared by all multi-class reductions (oaa, csoaa, log_multi, ect, cbify,...), but multi-class log loss is well defined only when:
* loss_function=logistic is used (so, the raw_predictions can be interpreted as probability after applying logistic link function).
* We have predictions (probabilities) for all classes. This is needed because we need to normalize the probabilities, so they sum up to one (which is not guaranteed if you just apply `--oaa --raw_predictions` and logistic link function).

Based on these two constrains, multi-class log loss can be computed only for `--oaa` (and `--csoaa_ldf=mc`, see below). So my current plan is to implement first an option `--probabilities` (which would be compatible with `--oaa` and `--csoaa_ldf=mc` and maybe also binary classification). This would result in reporting (predicted conditional) probabilities of all classes in the file generated by `--predictions`. I think many users will find such option useful (see e.g. #724). The multi-class log loss will be reported in the final printout whenever `--probabilities` are requested.

Maybe we can add another option, which will change the progress reports from zero-one loss to logistic loss. Or should this be switched automatically when `--probabilities` are requested? I am not sure about this now (but it is not so important for me, once I see both losses in the final printout).

For reporting two losses in the final printout, I will need to add some global variables to [global_data.h](https://github.com/JohnLangford/vowpal_wabbit/blob/62fbc5553a1/vowpalwabbit/global_data.h#L203-L221), we already have 6 loss variables there, but that should not be a problem.

It seems it makes no sense to report multi-class log loss for `--csoaa`, where the costs can be any real numbers, so there is no clear way how to convert the predicted costs to class probabilities. However, `--csoaa_ldf=mc loss_function=logistic` can used for multi-class problems where label-dependent features are needed and reporting multi-class log loss would be very helpful there (at least for me).

I am not sure whether to forbid `--probabilities` with e.g. `--loss_function=hinge`. In theory, there is no clear way how to convert hinge-optimized raw predictions to probabilities (cf. [Platt scaling](https://en.wikipedia.org/wiki/Platt_scaling)). But if we allow it (and use logistic link function and normalization on the hinge-optimized raw predictions), anyone can at least test how it influences the multi-class log loss (based on my experience, it is worse than when using `--loss_function=logistic`, even in cases when it has better accuracy).
